### PR TITLE
(MODULES-11321) Use new Puppet http runtime; require Puppet 6.11 or newer

### PR DESCRIPTION
**Pull request by tvpartytonight**
This PR changes the puppet http components and switches it to use the newer
http `runtime`. It also refactors the unit tests to stand up webbrick server for
easier to understand unit testing. Finally, it updates the version of vault used
in testing to the current version.

### CODE REVIEW

File: spec/acceptance/nodesets/docker/PuppetserverDockerfile
Comment (tvpartytonight) [inline spec/acceptance/nodesets/docker/PuppetserverDockerfile:3]: I'm not sure how the tests set up 6/7.x branches for the spec testing, but locally testing on edge works for me.

**Comment by bastelfreak** [comment]
@tvpartytonight is this still WIP or ready for review/merge?

**Comment by tvpartytonight** [comment]
@bastelfreak I still have some things for this PR I'd like to get in but I don't think they are blocking merge; one thing that I would like to do is move this to testing the release that should go out today instead of [edge](https://github.com/voxpupuli/puppet-vault_lookup/pull/50#discussion_r883867363).

Beyond that, this is only tested in ruby 3.0, while I think it would be good to continue to have ruby 2.7 testing as well. Additionally, the testing matrix runs the same set of tests for Centos7/Deb8/Ubuntu16 but I don't think that categorization makes any difference since these tests run only with a docker container that I don't think utilizes anything with that matrix.

File: lib/puppet/functions/vault_lookup/lookup.rb
Comment (joshcooper) [inline lib/puppet/functions/vault_lookup/lookup.rb:69]: By default `vault_cert_path_segment` ends with '/', so the above concatenation works as expected:

```ruby
irb(main):016:0> vault_url + 'v1/auth/cert/' + 'login'
=> #<URI::HTTPS https://vault.example.com/v1/auth/cert/login>
```

But if the trailing slash is omitted, the results are surprising:

```
irb(main):017:0> vault_url + 'v1/auth/cert' + 'login'
=> #<URI::HTTPS https://vault.example.com/v1/auth/login>
```

Maybe append the trailing slash if it's missing?

File: lib/puppet/functions/vault_lookup/lookup.rb
Comment (tvpartytonight) [reply lib/puppet/functions/vault_lookup/lookup.rb:48]: Since vault namespacing is an enterprise feature and we don't have a good way to stand up a Vault Enterprise instance to ensure this works I am going to remove this for now.

File: lib/puppet/functions/vault_lookup/lookup.rb
Comment (joshcooper) [inline lib/puppet/functions/vault_lookup/lookup.rb:5]: The `optional_param` vs `undef` thing has tripped me up more than once... There's more history in  [PUP-8101](https://tickets.puppetlabs.com/browse/PUP-8101?focusedCommentId=502043&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-502043) and [PUP-11066](https://tickets.puppetlabs.com/browse/PUP-11066?focusedCommentId=827195&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-827195)

My understanding is `optional_param` says I can accept this positional argument or not. But it doesn't specify the data type of the argument. Using `Variant` above will work or you can use a shorter form:

```
optional_param 'Optional[String]', :vault_cert_path_segment
````

This is because the [`undef` data type](https://puppet.com/docs/puppet/7/lang_data_undef.html) is defined so that:

> Several abstract data types can match the undef value ... The Optional type ... matches undef in addition to that type

So this returns `true`:

```
$ bx puppet apply -e 'notice(undef =~ Optional[String])'                                
Notice: Scope(Class[main]): true
```

But this doesn't:

```
$ bx puppet apply -e 'notice(42 =~ Optional[String])'
Notice: Scope(Class[main]): false
```

File: lib/puppet/functions/vault_lookup/lookup.rb
Comment (joshcooper) [inline lib/puppet/functions/vault_lookup/lookup.rb:7]: Should we declare `return_type`?

Comment (tvpartytonight) [reply lib/puppet/functions/vault_lookup/lookup.rb:7]: yeah, that seems appropriate.

**Comment by tvpartytonight** [comment]
@bastelfreak I think this is ready for merge, please let me know if there are additional changes required before merge.

**Review by joshcooper — APPROVED** [review]

**Comment by bastelfreak** [comment]
Thanks for the awesome work!

**Pull request merged**
